### ISSUE TITLE
Way to disable image/container GC

### ISSUE THREAD

**Issue by sorah** [GoogleCloudPlatform/kubernetes#3393]
#3097 added support for garbage collection of images. Sometime I run `docker pull` or `docker build` by hand on `kubelet` host. I guess garbage collection is taking things wrong way sometimes... (e.g. built image have gone immediately)

Are there any possibility to provide a way disabling image/container GC permanently or temporarily?

**Comment by sorah** [comment]
(If there's possibility, I'll try to write a patch)

**Comment by thockin** [comment]
There is no way currently.  I would be open to ideas, though we did just
change the GC TTL to 1 minute, so "instant delete" should not happen.

**Comment by sorah** [comment]
But it's happening on my environment. Initial pull won't succeed; May be my machine is slow, but it's another problem.

**Comment by thockin** [comment]
What dies the log say?

**Comment by brendandburns** [comment]
Yeah, gc will def. conflict with manual pull/build on a worker node.

It would be easy enough to add a flag to disable gc in the kubelet.

Dawn is in the midst of working on the gc code and we may be able to
restrict gc to only images that we pull by adding our own k8s tags to the
images.  That's prob. the correct strategy long term.

Ultimately, it would be way better if docker just handled gc, since it has
much better info for making the gc decision.

Brendan

**Comment by thockin** [comment]
I was thinking for short term something like a lock file that admin could
touch and which kubelet would respect and skip GC while it exists.

**Comment by sorah** [comment]
here's log. Like this during `pull`. Same environment without kubelet succeeds w/o error.

>FATA[0131] Error pulling image (base) from XXXX, Driver devicemapper failed to get image parent XXXX: Error activating devmapper device for 'XXXX': Error running DeviceCreate (ActivateDevice) dm_task_run failed

and, just, "no image exists" after an build.

> restrict gc to only images that we pull by adding our own k8s tags

I agree.

>I was thinking for short term something like a lock file that admin could
>touch and which kubelet would respect and skip GC while it exists.

I'm thinking something like that too. Lock should exist on etcd or local FS, which's better?

**Comment by thockin** [comment]
Does it fail for all images?  Or just this one?

If we have a GC lockout, it has to be local.  There is no api-stored config
object and we are moving away from direct etcd access.

**Comment by sorah** [comment]
Fails on another images too.

**Comment by thockin** [comment]
ALL images or just some?  Even kubernetes/pause?  Any pattern?

**Comment by sorah** [comment]
kubernetes/pause, It's already running, so Docker doesn't have to pull again and it won't be GCed. `docker run` succeeds.

Got another error for `centos:7`.

```
$ sudo docker run --rm -ti centos:7 bash
Unable to find image 'centos:7' locally
centos:7: The image you are pulling has been verified
5b12ef8fd570: Pull complete
5b12ef8fd570: Download complete
8efe422e6104: Download complete
511136ea3c5a: Download complete
Status: Downloaded newer image for centos:7
FATA[0064] Error response from daemon: No such image: centos:7 (tag: 7)
```

**Comment by thockin** [comment]
So you ate seeing errors from plain old docker, too?

**Comment by sorah** [comment]
Sorry I couldn't get what 'plain old docker' meant. my original image and some other images can successfully pull+run w/o kubelet. The previous post's log is happened during kubelet is alive.

I tested again `docker pull`, `docker run` for `centos:7` image and my original image on same host, after shutting kubelet down. It was succeeded.

**Comment by thockin** [comment]
Ahh, well, can you test with the --minimum_container_ttl_duration set to
"1m"?

**Comment by dchen1107** [comment]
1) Yes, current image GC is conflict with "docker pull" and "docker build". Especially with latest docker, docker build cannot have tag assigned initially. 

2) @thockin I don't think --minimum_container_ttl_duration works for images. Current image GC is completely based on docker's "dangling=true" feature to get unused images, which returns only images without tags. This means any image are downloaded without tag has a potential to be removed before it is used; and any image with tag is not going to be removed even no one is referring to it for a long long time. Anyway I re-open the original GC issue, and plan to re-do this.

**Comment by goltermann** [comment]
If you have a patch for this, we'd be happy to review/accept it.

**Comment by sorah** [comment]
Hm, I'll try to write a patch. My plan is using a lock file to disable GC temporarily.

the root problem is as @brendandburns said:

>Dawn is in the midst of working on the gc code and we may be able to
>restrict gc to only images that we pull by adding our own k8s tags to the
>images.  That's prob. the correct strategy long term.

but this takes a time to fix, so for short-term fix, I prefer my plan.

**Comment by dchen1107** [comment]
The initial issue was fixed by #3642, and close this one.
